packages/core/src/DataSchema.ts

```diff
@@ -0,0 +1,173 @@
+                sumOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            } else if (updateBy === 'replace') {
+                if (bits !== 253 - this.config.REPL_NONCE_BITS)
+                    throw new Error(
+                        `Field must be ${
+                            253 - this.config.REPL_NONCE_BITS
+                        } bits`
+                    )
+                if (replOffset + bits > maxReplOffset) {
+                    throw new Error(
+                        `Invalid schema, field "${name}" exceeds available storage`
+                    )
+                }
+
+                const dataIndex =
+                    this.config.SUM_FIELD_COUNT + Math.floor(replOffset / 253)
+                const offset = replOffset % 253
+                replOffset += bits
+                return { ...field, dataIndex, offset, bits }
+            }
+            throw new Error(
+                `Invalid updateBy strategy for field ${name}: "${updateBy}"`
+            )
+        })
+    }
+
+    buildAttestation(change: { name: string; v: bigint }): Attestation {
+        const field: any = this.schema.find((f) => f.name === change.name)
+
+        if (field === undefined) {
+            throw new Error(`${change.name} not found`)
+        }
+        const fieldIndex: number = field.dataIndex
+        const x: bigint = change.v << BigInt(field.offset)
+
+        const maxVal: bigint = (BigInt(1) << BigInt(field.bits)) - BigInt(1)
+
+        if (x > maxVal << BigInt(field.offset)) {
+            throw new Error(`${change.name} exceeds allocated space`)
+        }
+
+        const attestation: Attestation = {
+            fieldIndex,
+            change: x,
+        }
+
+        return attestation
+    }
+
+    buildAttestations(changes: { name: string; v: bigint }[]): Attestation[] {
+        let attestations: Attestation[] = Array(this.schema.length).fill(null)
+        for (const change of changes) {
+            const field: any = this.schema.find((f) => f.name === change.name)
+
+            if (field === undefined) {
+                throw new Error(`${change.name} not found`)
+            }
+
+            const fieldIndex: number = field.dataIndex
+            let v: bigint = change.v << BigInt(field.offset)
+            const maxVal: bigint = (BigInt(1) << BigInt(field.bits)) - BigInt(1)
+
+            if (attestations[fieldIndex] !== null) {
+                v =
+                    change.v +
+                    (field.updateBy === 'sum'
+                        ? (attestations[fieldIndex].change >>
+                              BigInt(field.offset)) &
+                          maxVal
+                        : BigInt(0))
+            }
+
+            if (v > maxVal << BigInt(field.offset)) {
+                throw new Error(`${change.name} exceeds allocated space`)
+            }
+
+            if (attestations[fieldIndex] !== null)
+                v =
```

Comment:
This is very complicated, you can just do `v = v + attestations[fieldIndex].change`

Reply:
Unless it's a replacement field in which case it should be no operation (discard the previous attestations)

Reply:
Yep, I realize I was making it a bit more complicated than necessary by trying to overwrite the existing change value with bit shifts instead of just adding to it.